src/Modules/UngradedModules.jl

```diff
@@ -0,0 +1,3808 @@
+  end=#
+
+  return SQ,injection,projection
+end
+
+######################################
+# Not only for testing
+######################################
+@doc Markdown.doc"""
+  matrix_to_map(F::FreeMod{T}, A::MatElem{T}) where T
+Converts a given n×m-matrix into the corresponding morphism $A : R^n → F$, 
+with `rank(F) == m`.
+"""
+function matrix_to_map(F::FreeMod{T}, A::MatElem{T}) where T
+  R = base_ring(F)
+  F_domain = FreeMod(R, nrows(A))
+
+  phi = FreeModuleHom(F_domain, F, A)
+  return phi
+end
+
+@doc Markdown.doc"""
+  matrix_to_map(A::MatElem)
+Converts a given n×m-matrix into the corresponding morphism $A : R^n → R^m$.
+"""
+function matrix_to_map(A::MatElem)
+  R = base_ring(A)
+  F_codomain = FreeMod(R, ncols(A))
+  return matrix_to_map(F_codomain,A)
+end
+
+@doc Markdown.doc"""
+  isinjective(f::ModuleMap)
+Test if `f` is injective.
+"""
+function isinjective(f::ModuleMap)
+  return iszero(kernel(f)[1])
+end
+
+@doc Markdown.doc"""
+  issurjective(f::ModuleMap)
+Test if `f` is surjective.
+"""
+function issurjective(f::ModuleMap)
+  return image(f)[1] == codomain(f)
+end
+
+@doc Markdown.doc"""
+  isbijective(f::ModuleMap)
+Test if `f` is bijective.
+"""
+function isbijective(f::ModuleMap)
+  return isinjective(f) && issurjective(f)
+end
+
+@doc Markdown.doc"""
+    iswelldefined(H::ModuleMap)
+Test if `H` is well-defined.
+"""
+function iswelldefined(H::ModuleMap)
+  if typeof(H) <: FreeModuleHom
+    return true
+  end
+  M = domain(H)
+  C = matrix(present_as_cokernel(M).quo)
+  n,m = size(C)
+  g = M[1]
+  ImH = map(x -> H(x), gens(M))
+  for i=1:n
+    if !iszero(Base.sum([C[i,j]*ImH[j] for j=1:m]))
+      return false
+    end
+  end
+  return true
+end
+
+#############################################
+# Test hom
+#############################################
+function to_julia_matrix(A::Union{MatElem})
+  return eltype(A)[A[i, j] for i = 1:nrows(A), j = 1:ncols(A)]
+end
+
+function Base.reshape(M::MatElem, n, m)
+  julia_matrix = to_julia_matrix(M)
+  julia_matrix = reshape(julia_matrix, n, m)
+  R = base_ring(M)
+  mat_space = MatrixSpace(R, n, m)
+  return mat_space(julia_matrix)
+end
+
+function hom2(f1::MatElem{T}, g1::MatElem{T}) where T
+  R = base_ring(f1)
+  s1, s0 = size(f1)
+  t1, t0 = size(g1)
+
+  g2 = matrix_kernel(g1)
+  n = s1*t0
+  m = s0*t0 + s1*t1
+  delta::MatrixElem{T} = zero_matrix(R, m,n)
+  #delta::Matrix{T} = zeros(R, n, m)
+  for j=1:m
+    #b_vector::Vector{T} = [i == j ? R(1) : R(0) for i=1:m]
+    b_vector::MatrixElem{T} = zero_matrix(R, 1,m)
+    b_vector[1,j] = R(1)
+    #println(typeof(Core.Array(b_vector[1,1:s0*t0])))
+    A = reshape(b_vector[1,1:s0*t0], s0, t0)
+    #Mat = AbstractAlgebra.MatrixSpace(R, t0, s0)
+    #A = Mat(A)
+    B = reshape(b_vector[1,s0*t0+1:length(b_vector)], s1, t1)
+    #Mat = AbstractAlgebra.MatrixSpace(R, t1, s1)
+    #B = Mat(B)
+    res = f1*A - B*g1
+    res = reshape(res, 1, length(res))
+    for i=1:length(res)
+      delta[j,i] = res[1,i]
+    end
```

Comment:
It may be possible to replace this by the following (assuming `length(res)` is `n` ?
```suggestion
    delta[j,:] = reshape(res, n)
```